ebean-migration/src/main/java/io/ebean/migration/MigrationConfig.java

```diff
@@ -611,25 +652,14 @@ public void setFastMode(boolean fastMode) {
   }
 
   /**
-   * Default factory. Uses the migration's class loader and injects the config if necessary.
-   *
-   * @author Roland Praml, FOCONIS AG
+   * Default implementation for service-loader. Note: ServiceLoader is not thread safe,
+   * so it is better to retrun a new iterator each time.
    */
-  public class DefaultMigrationFactory implements JdbcMigrationFactory {
+  private class DefaultMigrationFactory implements Iterable<JdbcMigration> {
 
     @Override
-    public JdbcMigration createInstance(String className) {
-      try {
-        Class<?> clazz = Class.forName(className, true, MigrationConfig.this.getClassLoader());
-        JdbcMigration migration = (JdbcMigration) clazz.getDeclaredConstructor().newInstance();
-        if (migration instanceof ConfigurationAware) {
-          ((ConfigurationAware) migration).setMigrationConfig(MigrationConfig.this);
-        }
-        return migration;
-      } catch (Exception e) {
-        throw new IllegalArgumentException(className + " is not a valid JdbcMigration", e);
-      }
+    public Iterator<JdbcMigration> iterator() {
+      return ServiceLoader.load(JdbcMigration.class, getClassLoader()).iterator();
```

Comment:
I'm not sure about service loading JdbcMigration (each migration).  I have in my mind that it would be better to add 1 level of indirection (a JdbcMigrationSupplier or JdbcMigrationFactory) ... then typically only that 1 supplier/factory would need to be registered for service loading rather than each JdbcMigration.

Hmmm.